### ISSUE TITLE
Import Ensembl databases Error

### ISSUE THREAD

**Issue by Bobo22331** [genomehubs/easy-mirror-docker#1]
When I run the following command to import Ensembl databases, it occurs error like this:
Command:
```
docker run --rm \
             --name genomehubs-ensembl \
             --network genomehubs-network \
             -v ~/genomehubs/v1/ensembl/conf:/ensembl/conf:ro \
             genomehubs/easy-mirror:19.05 /ensembl/scripts/database.sh /ensembl/conf/database.ini
```
Error:
```
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
/ensembl/scripts/database.sh: line 14: [: =: unary operator expected
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
awk: cannot open /ensembl/conf/database.ini (No such file or directory)
ERROR: Unable to connect to mysql server as root.
       Check connection settings in /ensembl/conf/database.ini
```
So, is there something bug with the Dockerfile?

**Comment by rjchallis** [comment]
It looks like the problem here is that it can't find the INI configuration file for the database import. This should be mounted from `~/genomehubs/v1/ensembl/conf`.

Can you check that you have a `database.ini` file in this directory?

**Comment by Bobo22331** [comment]
But the `database.ini` file is just in the `~/genomehubs/v1/ensembl/conf`.

**Comment by rjchallis** [comment]
I'm not sure what is happening here, if awk cannot open the file then it seems like it must be a problem with mounting the `/ensembl/conf` directory. Could you try a command like:

```
ls -al ~/genomehubs/v1/ensembl/conf
```

to check the the file and directory permissions make sense and 

```
docker run --rm \
                   --name filetest \
                   -v ~/genomehubs/v1/ensembl/conf:/ensembl/conf:ro \
                   genomehubs/easy-mirror:19.05 ls -l /ensembl/conf/database.ini
```

to check whether the file is visible inside the container.

**Comment by Bobo22331** [comment]
The permission of `~/genomehubs/v1/ensembl/conf` like this:
```
total 8
drwxr-xr-x 1 root root 4096 Dec  3 17:54 .
drwxr-xr-x 1 root root 4096 Dec  1 15:38 ..
-rwxr-xr-x 1 root root  960 Dec  3 17:54 database.ini
-rw-r--r-- 1 root root 1454 Dec  1 15:38 setup.ini
```
But the file maybe not visible in the docker.
```
ls: cannot access '/ensembl/conf/database.ini': No such file or directory
```

**Comment by rjchallis** [comment]
I think this must be a permissions issue, then. The only ways I can reproduce it is to mount a directory that I don't have read access to, or to mount a directory that doesn't exist (by mis-spelling the path), both of which give a `No such file or directory` message from inside the container like you are getting. What I don't understand is that you are able to `ls` the directory before it is mounted so you clearly do have read permissions outside the container.

I note that the files are owned by root, but processes inside the container run with a UID of 1000. you can run processes as a different user by adding `-u 1000:1000` to the docker run command to see if this helps. Generally it is best not to run docker containers as root so probably better to try setting this up as and making sure files are owned by a non-privileged user, rather than forcing the container to run as root. I usually run this on servers where I have UID 1000 but it should also be fine with higher UID numbers.

Depending how your system is configured it is possible that there is also user namespace remapping to consider (but only likely if you are running this on a shared system managed by someone else).

It may be worth trying to mount a different directory to see if you can access it from inside the container (e.g. making a directory in `/tmp` to make sure there is no difficulty caused by user directory permissions.

**Comment by Bobo22331** [comment]
I have tried as you say, but it did not work!

**Comment by rjchallis** [comment]
OK, could you try this with a different container to see if the problem is specific to this container or a more general docker directory mounting issue:

```
docker run --rm \
           -v ~/genomehubs/v1/ensembl/conf:/ensembl/conf:ro \
           ubuntu:latest ls -l /ensembl/conf/database.ini
```

**Comment by Bobo22331** [comment]
Yeah, I have tried as you said on two PCs, but they all do not work with the same error. So, maybe it's a more general docker directory mounting issue. I think I need to install the easy mirror on my local PC according to the Dockerfile. Thank you for your help!

**Comment by rjchallis** [comment]
I've only tested this on linux. As you mention PCs, I wonder if this Windows-related behaviour may be relevant: [github.com/docker/machine/issues/2253](https://github.com/docker/machine/issues/2253). Seems like it has been causing problems with docker mounts for a few years.
